client/src/main/scala/io/delta/sharing/client/DeltaSharingClient.scala

```diff
@@ -888,27 +927,64 @@ class DeltaSharingRestClient(
     (version, respondedFormat, lines, queryIdOpt)
   }
 
-
-  private def getNDJson[T: Manifest](target: String, data: T): (Long, String, Seq[String]) = {
+  private def getNDJson[T: Manifest](target: String, data: T): ParsedDeltaSharingResponse = {
     val httpPost = new HttpPost(target)
     val json = JsonUtils.toJson(data)
     httpPost.setHeader("Content-type", "application/json")
     httpPost.setEntity(new StringEntity(json, UTF_8))
     val (version, capabilities, lines) = getResponse(httpPost)
-    (
-      version.getOrElse {
+    val (respondedFormat, includedEndStreamAction) = getRespondedHeaders(capabilities)
+
+    val response = ParsedDeltaSharingResponse(
+      version = version.getOrElse {
         throw new IllegalStateException("Cannot find Delta-Table-Version in the header")
       },
-      getRespondedFormat(capabilities),
-      lines
+      respondedFormat = respondedFormat,
+      includedEndStreamAction = includedEndStreamAction,
+      lines,
+      capabilities = capabilities
     )
+    checkEndStreamAction(response)
+    response
   }
 
-  private def getRespondedFormat(capabilities: Option[String]): String = {
-    val capabilitiesMap = getDeltaSharingCapabilitiesMap(capabilities)
-    capabilitiesMap.get(RESPONSE_FORMAT).getOrElse(RESPONSE_FORMAT_PARQUET)
+  private def checkEndStreamAction(response: ParsedDeltaSharingResponse): Unit = {
+    if (includeEndStreamAction) {
+      // Only perform additional check when includeEndStreamAction = true
+      response.includedEndStreamAction match {
+        case Some(true) =>
+          val lastLineAction = JsonUtils.fromJson[SingleAction](response.lines.last)
+          if (lastLineAction.endStreamAction == null) {
+            throw new IllegalStateException("Client sets includeEndStreamAction=true in the " +
+              s"header, server responded with the header set to true(${response.capabilities}, " +
+              s"and ${response.lines.size} lines, and last line parsed as " +
+              s"${lastLineAction.unwrap.getClass()}, for query($dsQueryId).")
+          }
+        case Some(false) =>
+          throw new IllegalStateException(
+            "Client sets includeEndStreamAction=true in the header, but server responded with" +
+              s" the header set to false(${response.capabilities}), for query($dsQueryId)."
```

Comment:
We should be more tolerant and warn here as well.
An non-DB server may read our spec, and decide to not implement this but send back false anyways to tell the client that it is not sending endAction.